locator/network_topology_strategy.cc

```diff
@@ -336,40 +337,177 @@ future<tablet_map> network_topology_strategy::allocate_tablets_for_new_table(sch
+        }
+        for (const auto& node : nodes) {
+            const auto& host_id = node->host_id();
+            if (!existing.contains(host_id)) {
+                candidate.nodes.emplace_back(host_id, load.get_load(host_id));
             }
         }
+        if (candidate.nodes.empty()) {
+            candidate_racks.pop_back();
+            tablet_logger.trace("allocate_replica: no candidate nodes left on rack={}", rack);
+            // Note that this rack can't be in new_racks since
+            // those had no existing replicas and if current rack has no nodes
+            // we skip it in the beginning of the loop body
+            continue;
+        }
+        // Sort candidate nodes in each rack in descending load order
+        // so we want to allocate first from the least loaded nodes.
+        // Do shuffle + stable_sort to shuffle nodes with equal load.
+        std::shuffle(candidate.nodes.begin(), candidate.nodes.end(), rnd_engine);
+        std::stable_sort(candidate.nodes.begin(), candidate.nodes.end(), node_load_cmp);
+    }
+
+    if (candidate_racks.empty()) {
+        on_internal_error(tablet_logger, format("No candidate racks found for dc={} allocated={} rf={}: existing={}", dc, dc_node_count, dc_rf, replicas_per_rack));
+    }
 
-        tablet_replica_set replicas;
-        for (auto&& ep : tracker.replicas()) {
-            replicas.emplace_back(tablet_replica{ep, load.next_shard(ep)});
+    // ensure fairness across racks (in particular if rf < number_of_racks)
+    candidates_list::iterator candidate_rack = new_racks.empty()
+            ? candidate_racks.begin() + (tb.id % candidate_racks.size())
+            : new_racks[tb.id % new_racks.size()];
+
+    auto allocate_replica = [&] (candidates_list::iterator& candidate) {
+        const auto& rack = candidate->rack;
+        auto& nodes = candidate->nodes;
+        if (nodes.empty()) {
+            on_internal_error(tablet_logger, format("Candidates vector for rack={} is empty for allocating tablet replicas in dc={} allocated={} rf={}", rack, dc, dc_node_count, dc_rf));
+        }
+        auto host_id = nodes.back().host;
+        auto replica = tablet_replica{host_id, load.next_shard(host_id)};
+        const auto& node = topo.get_node(host_id);
+        auto inserted = replicas_per_rack[node.dc_rack().rack].insert(host_id).second;
+        // Sanity check that a node is not used more than once
+        if (!inserted) {
+            on_internal_error(tablet_logger, format("allocated replica={} node already used when allocating tablet replicas in dc={} allocated={} rf={}: replicas={}",
+                    replica, dc, dc_node_count, dc_rf, replicas));
+        }
+        nodes.pop_back();
+        tablet_logger.trace("Allocated tablet replica={} dc={} rack={}: nodes remaining in rack={}", replica, node.dc_rack().dc, node.dc_rack().rack, nodes.size());
+        if (nodes.empty()) {
+            candidate = candidate_racks.erase(candidate);
```

Comment:
shouldn't we consider new_racks here and below? There may be new_racks left, but this may jump to a non-new rack.

Maybe new_racks should be a separate list, and first we try to allocate from it, and the from the non-new list.

Reply:
done in v4